File: src/day12.rs
/// Day 12: Subterranean Sustainability
/// 
/// Figure out which plants are growing based on the plants around them

use std::collections::HashMap;

pub struct Plants {
    pub plants: Vec<char>,
    pub rules: HashMap<Vec<char>, char>,
    pub zero: isize
}

impl Plants {
    pub fn new(initial: &str, rules: &str) -> Self {
        let mut plants = Self { 
            plants: initial.chars().collect(),
            rules: HashMap::new(),
            zero: 0,
        };
        for line in rules.lines() {
            let mut parts = line.split(" => ");
            let input = parts.next().unwrap();
            let output = parts.next().unwrap();
            plants.rules.insert(input.chars().collect(), output.chars().next().unwrap());
        }
        plants
    }

    pub fn iterate(&mut self, times: usize) {
        for t in 0..times {
            if self.plants.iter().take(5).any(|c| *c == '#') {
                for _i in 0..5 { self.plants.insert(0, '.'); }
                self.zero += 5;
            }
            if self.plants.iter().rev().take(5).any(|c| *c == '#') {
                for _i in 0..5 { self.plants.push('.'); }
            }
            let mut new_plants = self.plants.clone();
            for i in 2..(self.plants.len() - 2) {
                let neighborhood = &self.plants[i - 2..=i+2];
                new_plants[i] = *self.rules.get(neighborhood).unwrap_or(&'.');
            }
            self.plants = new_plants;
        }
    }

    pub fn live_count(&self) -> isize {
        self.plants.iter().enumerate()
            .filter(|(_index, plant)| **plant == '#')
            .map(|(index, _plant)|{
                (index as isize) - self.zero
            })
            .sum()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_part_one() {
        let mut plants = Plants::new("#..#.#..##......###...###", "...## => #
..#.. => #
.#... => #
.#.#. => #
.#.## => #
.##.. => #
.#### => #
#.#.# => #
#.### => #
##.#. => #
##.## => #
###.. => #
###.# => #
####. => #");
        plants.iterate(20);
        assert_eq!(325, plants.live_count());
    }
}

File: python/src/day12.py
"""Day 12: Subterranean Sustainability

Conway's game of life with plants!  See which plants will live or die
each generation.
"""

from collections import deque
from itertools import islice
from typing import List, Dict


class Plants(deque):
    """A row of plants centered around zero"""

    def __init__(self, *args):
        deque.__init__(self, *args)
        self.zero = 0

    def score(self) -> int:
        """Sums up all the indices (including negative) that have a plant"""
        return sum(i - self.zero for i, pot in enumerate(self) if pot == '#')

    def __getitem__(self, key) -> islice:
        """Allows slicing a deque"""
        if isinstance(key, slice):
            return islice(self, key.start, key.stop, key.step)
        else:
            return deque.__getitem__(self, key)

    def copy(self):
        result = Plants(self)
        result.zero = self.zero
        return result


def next_generation(plants: Plants, rules: Dict[str, str]) -> Plants:
    """Given a row of pots with/without plants and some rules, creates
    the next generation of plants.

    The only rules that could increase the length in either direction
    require 3 dots on the end to check, so this makes sure there are 3
    dots on each end just in case.
    """
    this_generation = plants.copy()
    if any(c == '#' for c in this_generation[:3]):
        this_generation.extendleft(['.']*3)
        this_generation.zero += 3
    if any(c == '#' for c in this_generation[len(this_generation) - 3:]):
        this_generation.extend(['.']*3)

    next_generation = this_generation.copy()
    for i in range(2, len(this_generation) - 2):
        next_generation[i] = rules.get("".join(this_generation[i-2:i+3]), '.')
    return next_generation


def parse_rules(text: str) -> Dict[str, str]:
    """Parses the conversion rules from a block of text"""
    rules = {}
    for line in text.splitlines():
        pattern, _, output = line.partition(" => ")
        rules[pattern] = output
    return rules


def age(plants: Plants, generations: int, rules: Dict[str, str]) -> Plants:
    """Ages a set of plants n generations"""
    for i in range(generations):
        plants = next_generation(plants, rules)
    return plants


if __name__ == "__main__":
    with open("python/data/day12.txt", "r") as f:
        rules = parse_rules(f.read())
    initial_plants = Plants(
        "##..#..##....#..#..#..##.#.###.######..#..###.#.#..##.###.#.##..###..#.#..#.##.##..###.#.#...#.##..")

    # Part 1
    plants = age(initial_plants, 20, rules)
    print(plants.score())

    # Part 2: Wherein it's #@%!# linear above about 300

    plants = age(initial_plants, 300, rules)
    print(plants.score() + (50000000000 - 300)*86)
